// react-navigation/lib-rn/views/TabView/TabView.js
import React, { PureComponent } from 'react';
import { StyleSheet } from 'react-native';
import { TabViewAnimated, TabViewPagerPan } from 'react-native-tab-view';
import SceneView from '../SceneView';
import withCachedChildNavigation from '../../withCachedChildNavigation';

var babelPluginFlowReactPropTypes_proptype_NavigationTabScreenOptions = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationTabScreenOptions || require('react').PropTypes.any;

var babelPluginFlowReactPropTypes_proptype_NavigationRouter = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationRouter || require('react').PropTypes.any;

var babelPluginFlowReactPropTypes_proptype_NavigationState = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationState || require('react').PropTypes.any;

var babelPluginFlowReactPropTypes_proptype_NavigationAction = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationAction || require('react').PropTypes.any;

var babelPluginFlowReactPropTypes_proptype_NavigationRoute = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationRoute || require('react').PropTypes.any;

var babelPluginFlowReactPropTypes_proptype_NavigationScreenProp = require('../../TypeDefinition').babelPluginFlowReactPropTypes_proptype_NavigationScreenProp || require('react').PropTypes.any;

if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_TabViewConfig', {
  value: require('react').PropTypes.shape({
    tabBarComponent: require('prop-types').any,
    tabBarPosition: require('prop-types').oneOf(['top', 'bottom']),
    tabBarOptions: require('prop-types').shape({}),
    swipeEnabled: require('prop-types').bool,
    animationEnabled: require('prop-types').bool,
    lazy: require('prop-types').bool
  })
});
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_TabScene', {
  value: require('react').PropTypes.shape({
    route: babelPluginFlowReactPropTypes_proptype_NavigationRoute,
    focused: require('prop-types').bool.isRequired,
    index: require('prop-types').number.isRequired,
    tintColor: require('prop-types').string
  })
});


class TabView extends PureComponent {

  _handlePageChanged = index => {
    const { navigation } = this.props;
    navigation.navigate(navigation.state.routes[index].routeName);
  };

  _renderScene = ({ route }) => {
    const { screenProps } = this.props;
    const childNavigation = this.props.childNavigationProps[route.key];
    const TabComponent = this.props.router.getComponentForRouteName(route.routeName);
    return <SceneView screenProps={screenProps} component={TabComponent} navigation={childNavigation} />;
  };

  _getLabel = ({ route, tintColor, focused }) => {
    const options = this.props.router.getScreenOptions(this.props.childNavigationProps[route.key], this.props.screenProps || {});

    if (options.tabBarLabel) {
      return typeof options.tabBarLabel === 'function' ? options.tabBarLabel({ tintColor, focused }) : options.tabBarLabel;
    }

    if (typeof options.title === 'string') {
      return options.title;
    }

    return route.routeName;
  };

  _renderIcon = ({ focused, route, tintColor }) => {
    const options = this.props.router.getScreenOptions(this.props.childNavigationProps[route.key], this.props.screenProps || {});
    if (options.tabBarIcon) {
      return typeof options.tabBarIcon === 'function' ? options.tabBarIcon({ tintColor, focused }) : options.tabBarIcon;
    }
    return null;
  };

  _renderTabBar = props => {
    const {
      tabBarOptions,
      tabBarComponent: TabBarComponent,
      animationEnabled
    } = this.props;
    if (typeof TabBarComponent === 'undefined') {
      return null;
    }
    return <TabBarComponent {...props} {...tabBarOptions} navigation={this.props.navigation} getLabel={this._getLabel} renderIcon={this._renderIcon} animationEnabled={animationEnabled} />;
  };

  _renderPager = props => <TabViewPagerPan {...props} />;

  render() {
    const {
      router,
      tabBarComponent,
      tabBarPosition,
      animationEnabled,
      swipeEnabled,
      lazy,
      screenProps
    } = this.props;

    let renderHeader;
    let renderFooter;
    let renderPager;

    const { state } = this.props.navigation;
    const options = router.getScreenOptions(this.props.childNavigationProps[state.routes[state.index].key], screenProps || {});

    const tabBarVisible = options.tabBarVisible == null ? true : options.tabBarVisible;

    if (tabBarComponent !== undefined && tabBarVisible) {
      if (tabBarPosition === 'bottom') {
        renderFooter = this._renderTabBar;
      } else {
        renderHeader = this._renderTabBar;
      }
    }

    if (animationEnabled === false && swipeEnabled === false) {
      renderPager = this._renderPager;
    }

    const props = {
      lazy,
      animationEnabled,
      swipeEnabled,
      renderPager,
      renderHeader,
      renderFooter,
      renderScene: this._renderScene,
      onRequestChangeTab: this._handlePageChanged,
      navigationState: this.props.navigation.state,
      screenProps: this.props.screenProps,
      style: styles.container
    };

    return <TabViewAnimated {...props} />;
  }
}

TabView.propTypes = {
  tabBarComponent: require('prop-types').any,
  tabBarPosition: require('prop-types').oneOf(['top', 'bottom']),
  tabBarOptions: require('prop-types').shape({}),
  swipeEnabled: require('prop-types').bool,
  animationEnabled: require('prop-types').bool,
  lazy: require('prop-types').bool,
  screenProps: require('prop-types').shape({}),
  navigation: babelPluginFlowReactPropTypes_proptype_NavigationScreenProp,
  router: babelPluginFlowReactPropTypes_proptype_NavigationRouter,
  childNavigationProps: require('prop-types').shape({}).isRequired
};
TabView.propTypes = {
  tabBarComponent: require('prop-types').any,
  tabBarPosition: require('prop-types').oneOf(['top', 'bottom']),
  tabBarOptions: require('prop-types').shape({}),
  swipeEnabled: require('prop-types').bool,
  animationEnabled: require('prop-types').bool,
  lazy: require('prop-types').bool,
  screenProps: require('prop-types').shape({}),
  navigation: babelPluginFlowReactPropTypes_proptype_NavigationScreenProp,
  router: babelPluginFlowReactPropTypes_proptype_NavigationRouter,
  childNavigationProps: require('prop-types').shape({}).isRequired
};
const TabViewEnhanced = withCachedChildNavigation(TabView);

export default TabViewEnhanced;

const styles = StyleSheet.create({
  container: {
    flex: 1
  }
});
