// react-navigation/lib-rn/views/HeaderBackButton.js
import React from 'react';
import { I18nManager, Image, Text, View, Platform, StyleSheet } from 'react-native';

var babelPluginFlowReactPropTypes_proptype_LayoutEvent = require('../TypeDefinition').babelPluginFlowReactPropTypes_proptype_LayoutEvent || require('react').PropTypes.any;

import TouchableItem from './TouchableItem';

class HeaderBackButton extends React.PureComponent {
  static defaultProps = {
    pressColorAndroid: 'rgba(0, 0, 0, .32)',
    tintColor: Platform.select({
      ios: '#037aff'
    }),
    truncatedTitle: 'Back'
  };

  state = {};

  _onTextLayout = e => {
    if (this.state.initialTextWidth) {
      return;
    }
    this.setState({
      initialTextWidth: e.nativeEvent.layout.x + e.nativeEvent.layout.width
    });
  };

  render() {
    const {
      onPress,
      pressColorAndroid,
      width,
      title,
      tintColor,
      truncatedTitle
    } = this.props;

    const renderTruncated = this.state.initialTextWidth && width ? this.state.initialTextWidth > width : false;

    const backButtonTitle = renderTruncated ? truncatedTitle : title;

    // eslint-disable-next-line global-require
    const asset = require('./assets/back-icon.png');

    return <TouchableItem accessibilityComponentType="button" accessibilityLabel={backButtonTitle} accessibilityTraits="button" delayPressIn={0} onPress={onPress} pressColor={pressColorAndroid} style={styles.container} borderless>
        <View style={styles.container}>
          <Image style={[styles.icon, title && styles.iconWithTitle, { tintColor }]} source={asset} />
          {Platform.OS === 'ios' && title && <Text onLayout={this._onTextLayout} style={[styles.title, { color: tintColor }]} numberOfLines={1}>
              {backButtonTitle}
            </Text>}
        </View>
      </TouchableItem>;
  }
}

HeaderBackButton.propTypes = {
  onPress: require('prop-types').func,
  pressColorAndroid: require('prop-types').string,
  title: require('prop-types').string,
  tintColor: require('prop-types').string,
  truncatedTitle: require('prop-types').string,
  width: require('prop-types').number
};
const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: 'transparent'
  },
  title: {
    fontSize: 17,
    paddingRight: 10
  },
  icon: Platform.OS === 'ios' ? {
    height: 20,
    width: 12,
    marginLeft: 10,
    marginRight: 22,
    marginVertical: 12,
    resizeMode: 'contain',
    transform: [{ scaleX: I18nManager.isRTL ? -1 : 1 }]
  } : {
    height: 24,
    width: 24,
    margin: 16,
    resizeMode: 'contain',
    transform: [{ scaleX: I18nManager.isRTL ? -1 : 1 }]
  },
  iconWithTitle: Platform.OS === 'ios' ? {
    marginRight: 5
  } : {}
});

export default HeaderBackButton;
